Tests/TestingApps/PaywallsTester/PaywallsTester/Data/OfferingsPaywallsViewModel.swift

```diff
@@ -0,0 +1,182 @@
+//
+//  OfferingsPaywallsViewModel.swift
+//  PaywallsTester
+//
+//  Created by James Borthwick on 2024-04-25.
+//
+
+import Foundation
+import RevenueCat
+
+struct OfferingPaywall: Hashable {
+    let offering: OfferingsResponse.Offering
+    let paywall: PaywallsResponse.Paywall
+}
+
+struct PresentedPaywall: Hashable {
+    var offering: Offering
+    var mode: PaywallViewMode
+    var responseOfferingID: String
+}
+
+@Observable
+final class OfferingsPaywallsViewModel {
+
+    var apps: [DeveloperResponse.App]
+
+    init(apps: [DeveloperResponse.App]) {
+        self.apps = apps
+    }
+
+    var offeringsPaywalls: Result<[OfferingPaywall], NSError>? {
+        didSet {
+            Task { @MainActor in
+                refreshPresentedPaywall()
+            }
+        }
+    }
+
+    var presentedPaywall: PresentedPaywall?
+
+    @MainActor
+    func updateOfferingsAndPaywalls() async {
+        do {
+            let appCopy = apps
+            async let appOfferings = Self.fetchOfferings(for: appCopy).all
+            async let appPaywalls = Self.fetchPaywalls(for: appCopy).all
+
+            let offerings = try await appOfferings
+            let paywalls = try await appPaywalls
+
+            let offeringPaywallData = OfferingPaywallData(offerings: offerings, paywalls: paywalls)
+
+            self.offeringsPaywalls = .success(
+                offeringPaywallData.paywallsByOffering()
+            )
+
+        } catch let error as NSError {
+            self.offeringsPaywalls = .failure(error)
+            Self.logger.log(level: .error, "Could not fetch offerings/paywalls: \(error)")
+        }
+    }
+    
+    @MainActor
+    func getAndShowPaywallForID(id: String) async {
+
+        showPaywallForID(id)
+
+        // in case data has changed since last fetch
+        await updateOfferingsAndPaywalls()
+
+        showPaywallForID(id)
+    }
+
+    private static var logger = Logging.shared.logger(category: "Paywalls Tester")
+
+}
+
+// Private helpers
+extension OfferingsPaywallsViewModel {
```

Comment:
might as well make this a private extension and then the modifier isn't needed anymore

Reply:
Done!